apps_download/models/product_product.py

```diff
@@ -0,0 +1,112 @@
+
+    @api.multi
+    def create_dependency_list(self):
+        ret_val = {}
+
+        def child_dependency(children):
+            res = self.env['product.product']
+            for child in children:
+                if not child.dependent_product_ids:
+                    continue
+                res += child.dependent_product_ids
+                res += child_dependency(child.dependent_product_ids)
+            return res
+        for product in self:
+            ret_val[product.id] = product.dependent_product_ids
+            if product.dependent_product_ids:
+                ret_val[product.id] += child_dependency(
+                    product.dependent_product_ids)
+        return ret_val
+
+    @api.multi
+    def generate_zip_file(self):
+        for product in self:
+            if not product.module_path:
+                continue
+            tmpdir = tempfile.mkdtemp()
+            tmpdir_2 = tempfile.mkdtemp()
+            dependent_products = product.create_dependency_list()
+            dependent_products = dependent_products[product.id]
+            for dependent_product in dependent_products:
+                if not dependent_product.module_path:
+                    continue
+                subprocess.Popen(['cp', '-r', dependent_product.module_path,
+                                  tmpdir], stdout=subprocess.PIPE)
+            subprocess.Popen(['cp', '-r', product.module_path, tmpdir],
+                             stdout=subprocess.PIPE)
+            time_value = time.strftime(
+                '_%y%m%d_%H%M%S')
+
+            tmp_zip_file = os.path.join(tmpdir_2, product.name) + time_value
+            shutil.make_archive(tmp_zip_file, 'zip', tmpdir)
+            tmp_zip_file = '%s.zip' % tmp_zip_file
+            with open(tmp_zip_file, "rb") as file_obj:
+                try:
+                    data_encode = base64.encodestring(file_obj.read())
+                    self.env['ir.attachment'].create({
+                        'datas': data_encode,
+                        'datas_fname': tmp_zip_file,
+                        'type': 'binary',
+                        'name': product.name + time_value + '.zip',
+                        'res_model': product._name,
+                        'res_id': product.id,
+                        'product_downloadable': True,
+                    })
+                except:
+                    _logger.error('Error creating attachment %s' %
+                                  tmp_zip_file)
+            try:
+                shutil.rmtree(tmpdir)
+            except OSError as exc:
+                _logger.warning('Could not remove Tempdir %s, Errormsg %s' % (
+                    tmpdir, exc.message))
+            try:
+                shutil.rmtree(tmpdir_2)
+            except OSError as exc:
+                _logger.warning('Could not remove Tempdir 2 %s, Errormsg %s' % (
+                    tmpdir, exc.message))
```

Comment:
can we have the above cleanup in a `finally:` block?

Reply:
@gurneyalex  And what exactly goes to the finally: block? Let me know I will change accordingly.

Reply:
The exception handling is like this is also used in the core report.py, what can be done better? Shall we write a success log in finally, or what are you expecting there?

Reply:
@bizzappdev what I'm suggesting is having 

```
try:
   <body of the method>
finally:
   <the two try..except block in charge of calling shutil.rmtree>
```

This ensures that the cleanup is attempted whatever exception happens in the code inside the try: block.